externals/kyuubi-trino-engine/src/main/scala/org/apache/kyuubi/engine/trino/TrinoStatement.scala

```diff
@@ -0,0 +1,195 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.engine.trino
+
+import java.util.ArrayList
+import java.util.concurrent.ArrayBlockingQueue
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+import scala.concurrent.duration
+import scala.concurrent.duration.Duration
+import scala.util.control.Breaks._
+
+import com.google.common.base.Verify
+import io.trino.client.ClientSession
+import io.trino.client.Column
+import io.trino.client.StatementClient
+import io.trino.client.StatementClientFactory
+
+import org.apache.kyuubi.KyuubiSQLException
+import org.apache.kyuubi.engine.trino.TrinoStatement._
+
+/**
+ * Trino client communicate with trino cluster.
+ */
+class TrinoStatement(trinoContext: TrinoContext, sql: String) {
+
+  private lazy val trino = StatementClientFactory
+    .newStatementClient(trinoContext.httpClient, trinoContext.getClientSession, sql)
+
+  def getTrinoClient: StatementClient = trino
+
+  def getCurrentDatabase: String = trinoContext.getClientSession.getSchema
+
+  def getColumns: List[Column] = {
+    while (trino.isRunning) {
+      val results = trino.currentStatusInfo()
+      val columns = results.getColumns()
+      if (columns != null) {
+        return columns.asScala.toList
+      }
+      trino.advance()
+    }
+    Verify.verify(trino.isFinished())
+    val finalStatus = trino.finalStatusInfo()
+    if (finalStatus.getError == null) {
+      throw KyuubiSQLException(s"Query has no columns (#${finalStatus.getId})")
+    } else {
+      throw KyuubiSQLException(
+        s"Query failed (#${finalStatus.getId}): ${finalStatus.getError.getMessage}")
+    }
+  }
+
+  /**
+   * Execute sql and return ResultSet.
+   */
+  def execute(): Iterable[List[Any]] = {
+    val rowQueue = new ArrayBlockingQueue[List[Any]](MAX_QUEUED_ROWS)
+
+    val future = Future[Unit] {
+      while (trino.isRunning) {
+        val data = trino.currentData().getData()
+        if (data != null) {
+          data.asScala.map(_.asScala.toList)
+            .foreach(e => putOrThrow(rowQueue, e))
+        }
+        trino.advance()
+      }
+    }
```

Comment:
Any blocking operation in this block? IMO, `scala.concurrent.ExecutionContext.Implicits.global` is not a good choise for that. BTW, please give the variable a meaningful name rather than just `future`.

Reply:
> Any blocking operation in this block?

The resultSet have multiple batches, read and flush to buffer need blocking.

Reply:
Then let's use a dedicated `ExecutionContext` and make the thread size configurable rather than `scala.concurrent.ExecutionContext.Implicits.global`, the global one is something like global fork join pool, usually for noblocking calculation.

Reply:
Have addressed